sllm/serve/backends/transformers_backend.py

```diff
@@ -180,56 +215,199 @@ async def generate(self, request_data: Optional[Dict[str, Any]]):
-                "completion_tokens": completion_tokens,
-                "total_tokens": total_tokens,
-            },
-        }
+        if self.model is not None:
+            del self.model
 
-        return response
+    def stop(self) -> None:
+        """Wait for all requests to finish and shutdown the backend."""
+        with self.status_lock:
+            if self.status.value >= BackendStatus.STOPPING.value:
+                return
+            self.status = BackendStatus.STOPPING
+        while self.inf_status and len(self.inf_status.get()) > 0:
+            logger.info("Waiting for all requests to finish")
+            time.sleep(1)
+        logger.info("All requests finished. Shutting down the backend.")
+        self.shutdown()
+
+    def get_current_tokens(self):
+        """Return a list of all ongoing request tokens."""
+        with self.status_lock:
+            if self.status != BackendStatus.RUNNING:
+                return []
+
+        status = self.inf_status.get()
+        logger.info(f"Current tokens: {status}")
+        return status
+
+    def resume_kv_cache(self, request_datas):
+        logger.info(f"Resuming cache for {request_datas}")
+        with torch.no_grad():
+            input_ids = torch.tensor(request_datas).reshape(1, -1).to("cuda")
+            output = self.model.generate(
+                input_ids,
+                past_key_values=self.past_key_values,
+                max_new_tokens=1,
+                return_dict_in_generate=True,
+                return_legacy_cache=True,
+            )
+            self.past_key_values = output.past_key_values
+        logger.info(f"Resumed {len(self.past_key_values[0][0][0][0])} tokens")
+
+    def resume_generate(
+        self, request_data: Optional[Dict[str, Any]], current_output
+    ):
+        with self.status_lock:
+            if self.status != BackendStatus.RUNNING:
+                return {"error": "Model not initialized"}
+
+        assert self.model is not None
+
+        model_name = request_data.get("model", "dummy-model")
+        messages = request_data.get("messages", [])
+        temperature = request_data.get("temperature", 0.7)
+        max_tokens = request_data.get("max_tokens", 10)
+
+        # Combine messages to form the prompt
+        prompt = self.tokenizer.apply_chat_template(
+            messages, add_generation_prompt=True, tokenize=False
+        )
 
-    async def shutdown(self):
-        pass
+        if not prompt:
+            return {"error": "Missing prompt in request data"}
 
-    async def stop(self):
-        pass
+        inputs = self._tokenize(prompt)
+        prompt_tokens = inputs.input_ids.shape[1]
 
-    async def get_current_tokens(self):
-        logger.error("Not implemented")
-        raise NotImplementedError
+        # Generate response
+        try:
+            with torch.no_grad():
+                current_output = (
+                    torch.tensor(current_output).reshape(1, -1).to("cuda")
```

Comment:
This should be consistent with the model's device. In our testing, model is on cpu